bundle/src/projects/commercial/modules/elements-manager/targeting.ts

```diff
@@ -0,0 +1,146 @@
+
+		// Transform the serialized payload into an array of elements
+		const elements: GuElement[] = json.map((element) => ({
+			...element,
+			assets: element.assets.map((asset) => ({
+				...asset,
+				path: `${assetUrl}/${asset.path}`,
+			})),
+			targeting: element.targeting.map(({ key, value }) => ({
+				key,
+				value: new Set(value.split(',')),
+			})),
+		}));
+
+		return elements;
+	} catch (error) {
+		console.error(error);
+		return [];
+	}
+};
+
+/**
+ * Obtain consented or consentless page targeting object, based on consent state.
+ */
+export const getPageTargetingForElements = (
+	consentState: ConsentState,
+): PageTargeting => {
+	if (consentState.canTarget) {
+		return buildPageTargeting({
+			consentState,
+			adFree: commercialFeatures.adFree,
+			clientSideParticipations: getSynchronousParticipations(),
+		});
+	} else {
+		return buildPageTargetingConsentless(
+			consentState,
+			commercialFeatures.adFree,
+		);
+	}
+};
+
+/**
+ * Determines if a given targeting object matches a single condition
+ *
+ * @example
+ * satisfiesRule({ at: "fixed-puppies "}, { key: "at", value: new Set(["fixed-puppies", "fixed-cats"]) }); => true
+ */
+const satisfiesRule = (
+	pageTargeting: PageTargeting,
+	condition: TargetingRule,
+): boolean => {
+	const targetingValue = pageTargeting[condition.key];
+	if (!targetingValue) {
+		// If there is no value in the targeting for the corresponding key then
+		// this rule can't be satisfied
+		return false;
+	} else if (isString(targetingValue)) {
+		// If there is a single value in the targeting for this key, check that
+		// string is present in the condition's values
+		return condition.value.has(targetingValue);
+	} else {
+		// If there's multiple values in the targeting for this key, check that
+		// at least one is present in the condition's values
+		return targetingValue.some((v) => condition.value.has(v));
+	}
+};
+
+/**
+ * Check that all of the conditions in the target are satisfied by the page targeting object.
+ *
+ * We're effectively taking the **AND** of all of the conditions
+ */
+const satisfiesTargeting = (
+	pageTargeting: PageTargeting,
+	targetingRules: TargetingRules,
+): boolean =>
+	targetingRules.every((rule) => satisfiesRule(pageTargeting, rule));
+
+/**
+ * Compute the set of eligible elements for a given page targeting object.
+ */
+const eligibleElements = (
+	pageTargeting: PageTargeting,
+	elements: GuElement[],
+) =>
+	elements.filter(
+		(element) =>
+			// Element has at least one linked asset
+			element.assets.length > 0 &&
+			// The page targeting object supplied satisfies the element's targeting rules
+			satisfiesTargeting(pageTargeting, element.targeting),
+	);
+
+const selectAtRandom = <T>(candidates: T[]) =>
+	// TODO remove this type assertion when we can enable --noUncheckedIndexedAccess compiler option
+	candidates[Math.floor(Math.random() * candidates.length)] as T | undefined;
+
+export const initTargeting = (
```

Comment:
Is there perhaps is a more descriptive name for what this function is doing? It looks like it's finding an asset to serve given a list of elements and based on targeting.

Reply:
Have removed this function now, so naming doesn't apply anymore.